textfrag/src/types/directive/fragdirective.rs

```diff
@@ -0,0 +1,336 @@
+/// Fragment Directive object is collection of text fragments in the URL's fragment
+/// :~: is the fragment directive delmiter, followed by list of `[TextDirective]`
+use std::{cell::RefCell, collections::HashMap};
+
+use html5ever::tokenizer::{BufferQueue, Tokenizer, TokenizerOpts};
+use url::Url;
+
+use crate::{
+    extract::FragmentDirectiveTokenizer,
+    types::{
+        FragmentDirectiveError, FragmentDirectiveStatus, TextDirective, TextDirectiveStatus,
+        TextFragmentError,
+    },
+};
+
+pub const FRAGMENT_DIRECTIVE_DELIMITER: &str = ":~:";
+
+/// Fragment Directive defines the base url and collection of Text Directives
+#[derive(Default, Clone, Debug)]
+pub struct FragmentDirective {
+    #[allow(dead_code)]
+    url: Option<Url>,
+    text_directives: RefCell<Vec<TextDirective>>,
+}
+
+impl FragmentDirective {
+    pub fn text_directives(&self) -> Vec<TextDirective> {
```

Comment:
Ideally, I'd like to have a comment for every method. That's because a human description can add additional context that a function header might not miss.

For example, in this case:

```suggestion
    /// Returns all the defined text directives
    ///
    /// Note that this is a `Vector` and not a `HashSet`, because the directives get returned
    /// in the order they were defined. The order matters because...
    pub fn text_directives(&self) -> Vec<TextDirective> {
```

This is just an example. I'm actually not sure if the order matters. If it doesn't then a `HashSet` would be the more precise type.

Reply:
order will not matter as each of the object - i'm checking the recommendation!

Reply:
on second thoughts, even though the order of the text directives does not matter, I don't think we'll have any major benefits in picking up HashSet over the vectors - below are my reasonings:
1. the number of text directives will be always limited - it shall not be more than a handful directives
2. tokenizer processes the directives in an iterative manner and since there is no lookup requirements in here the HashSet advantage of O(1) operation is not that significant

given these thoughts, I am inclined to retain the vector over the HashSet.

please do let me know if I am wrong in the above understanding!